File: Ruby/civitas/lib/tablero.rb
#encoding: utf­8

require_relative 'casilla'
require_relative 'casilla_juez'
module Civitas
  class Tablero
    attr_accessor :numCasillaCarcel, :porSalida, :tieneJuez, :casillas, :casilla
    
    def initialize(casillaCarcel)
        if(casillaCarcel < 1)
          @numCasillaCarcel = 1
        else
          @numCasillaCarcel=casillaCarcel          
        end
          @casillas=Array.new
          casilla = Casilla.new("Salida")
          @casillas << casilla
          @porSalida=0
          @tieneJuez=false
        

      end
      
      private
      def correcto1()
        if((@casillas.size() > @numCasillaCarcel) && @tieneJuez == true)
          return true
        else
          return false
        end
      end
      
      def correcto(numCasilla)
        if(correcto1() && numCasilla>=0 && numCasilla<20)
          return true
        else
          return false
        end
      end
      
      public
      def tamanio()
        return @casillas.size()
      end
      
      def get_casilla(numCasilla)
        if(correcto(numCasilla))
          return @casillas.at(numCasilla)
        else
          return nil
        end
      end
      
      def get_por_salida()
        if(@porSalida > 0)
          v1=@porSalida
          @porSalida=@porSalida-1
          return v1
        else
          return @porSalida
        end
      end
      
      def aniade_casilla(casilla)
        if(@casillas.size() == @numCasillaCarcel)
          casill = Casilla.new("Carcel")
          @casillas << casill
        end
        
        @casillas << casilla
      end
      
      def aniade_juez
        if(@tieneJuez == false)
          casilla = CasillaJuez.new(5, "Juez")
         @casillas << casilla
          @tieneJuez = true
        end
      end
      
      def nueva_posicion(actual, tirada)
        if(!correcto1())
          return -1
        else
          suma = actual+tirada
          sumaF=suma%20
          if(suma!=sumaF)
            @porSalida=@porSalida+1
            puts @porSalida
          end
        end
        return sumaF
      end
      
      def calcular_tirada(origen, destino)
        tirada = destino - origen
        if(tirada >=0)
          return tirada
        else
          tirada = tirada + 20
          return tirada
        end
      end
      
  end
end

File: Java/civitas/src/civitas/Tablero.java
package civitas;
import java.util.ArrayList;

public class Tablero {
    private int numCasillaCarcel;
    private ArrayList<Casilla> casillas = new ArrayList<>();
    private int porSalida;
    private boolean tieneJuez;
    
    
    Tablero(int casillaCarcel){
        if(casillaCarcel < 1)
            numCasillaCarcel = 1;
        else
            numCasillaCarcel = casillaCarcel;
        
        Casilla salida = new Casilla("Salida");
        casillas.add(salida);
        porSalida = 0;
        tieneJuez = false;
    }


    private boolean correcto(){
        if((casillas.size() > numCasillaCarcel) && tieneJuez == true){
            return true;
        }
        else{
            return false;
        }
            
    }
    
    private boolean correcto(int numCasilla){
        if(correcto() && numCasilla >= 0 && numCasilla < 20)
            return true;
        else
            return false;
    }
    int size(){
        return casillas.size();
    }
    public int getCarcel() {
        return numCasillaCarcel;
    }

    public Casilla getCasilla(int numCasilla) {
        if(correcto(numCasilla))
            return casillas.get(numCasilla);
        else
            return null;
    }

    public int getPorSalida() {
        if(porSalida > 0){
            int v1 = porSalida;
            porSalida--;
            return v1;
        }
        else
            return porSalida;
    }

    public boolean isTieneJuez() {
        return tieneJuez;
    }
    
    public void aniadeCasilla(Casilla casilla){
        if(casillas.size() == numCasillaCarcel){
            casillas.add(new Casilla("Carcel"));
            casillas.add(casilla);
        }
        else
            casillas.add(casilla);
                    
    }
    
    public void aniadeJuez(){
        if(tieneJuez == false){
            Casilla juez = new CasillaJuez(5, "Juez");
            casillas.add(juez);
            tieneJuez=true;
        }
    }
    
    public int nuevaPosicion(int actual, int tirada){
        if(!correcto())
            return -1;
        else{
            int suma = actual+tirada;
            int sumaF=suma%20;
            if(sumaF != suma){
                porSalida++;
                System.out.println(porSalida);
            }
            return sumaF;
        }
    }
    
    public int calcularTirada(int origen, int destino){
        int tirada = destino-origen;
        if(tirada >= 0)
            return tirada;
        else{
            tirada = tirada + 20;
            return tirada;
       }
            
                    
        
    }
   
}
